[WIP] Prototype of parallel actor.

jovany-wang opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number
#21547
<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

bveeramani commented:
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

jovany-wang commented:
I believe this is almost done. Will confirm and refine as soon as possible.

jovany-wang commented:
@kfstorm I'm looking forward your comments on the API designing and the code structures as the 1st step reviewing process.

kfstorm commented on .bazelrc:
```diff
@@ -187,6 +187,8 @@ build:ubsan --copt -g
 build:ubsan --linkopt -fsanitize=undefined
 build:ubsan --linkopt -fno-sanitize-recover=all
 
+build --jobs=4
```
Revert

kfstorm commented on java/api/src/main/java/io/ray/api/runtime/RayRuntime.java:
```diff
@@ -284,4 +285,7 @@
   /** Create runtime env instance at runtime. */
   RuntimeEnv createRuntimeEnv(Map<String, String> envVars);
+  
+  /// TODO(qwang): Move out from Ray.java
```
Is the TODO already done?

jovany-wang replied:
Yes.

kfstorm commented on java/test/src/main/java/io/ray/test/BaseTest.java:
```diff
@@ -11,6 +11,7 @@
   @BeforeMethod(alwaysRun = true)
   public void setUpBase(Method method) {
     Assert.assertFalse(Ray.isInitialized());
+//    System.setProperty("ray.run-mode", "SINGLE_PROCESS");
```
Remove

kfstorm commented on java/api/src/main/java/io/ray/api/utils/parallelactor/ParallelCall.java:
```diff
@@ -0,0 +1,666 @@
+  // ===========================
+  // Methods for actor creation.
+  // ===========================
+  public static <A> ParallelActorCreator<A> parallelActor(RayFunc0<A> f) {
```
I think `actor` is better than `parallelActor` because parallel is mentioned in the package name.

kfstorm commented on java/api/src/main/java/io/ray/api/utils/parallelactor/ParallelActor.java:
```diff
@@ -0,0 +1,46 @@
+    parallelExecutorHandle = handle;
+  }
+
+  public ParallelInstance<A> getParallel(int index) {
```
```suggestion
  public ParallelInstance<A> getInstance(int index) {
```

kfstorm commented on java/api/src/main/java/io/ray/api/utils/parallelactor/ParallelActor.java:
```diff
@@ -0,0 +1,46 @@
+    return null;
+  }
+
+  public ActorHandle<? extends ParallelActorExecutor> getParallelExecutorHandle() {
```
```suggestion
  public ActorHandle<? extends ParallelActorExecutor> getExecutor() {
```

kfstorm commented on java/api/src/main/java/io/ray/api/utils/parallelactor/ParallelActor.java:
```diff
@@ -0,0 +1,46 @@
+    return parallelNum;
+  }
+
+  public int getNextIndex() {
```
Is this API needed? I think this API is not compatible with randomized strategies (if we will implement one later).

kfstorm replied:
If it's for internal use only, make drop `public`.

kfstorm replied:
And maybe the implementation should be in the runtime module?

jovany-wang replied:
Yes. This API is moved to the strategy implementation.

kfstorm commented on java/api/src/main/java/io/ray/api/utils/parallelactor/ParallelContext.java:
```diff
@@ -0,0 +1,12 @@
+
+  <A> ParallelActor<A> createParallelActorExecutor(ParallelStrategy strategy, int parallelNum, ConcurrencyGroup[] concurrencyGroups, RayFuncR<A> ctorFunc);
+
+  <R> ObjectRef<R> submitParallelTask(ParallelActor parallelActor, int parallelIndex, RayFuncR func, Object[] args);
```
```suggestion
  <R> ObjectRef<R> submitTask(ParallelActor parallelActor, int parallelIndex, RayFuncR func, Object[] args);
```

Because it only submits one task.

kfstorm commented on java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java:
```diff
@@ -288,6 +291,11 @@ public ConcurrencyGroup createConcurrencyGroup(
 
+  @Override
+  public ParallelContext getParallelContext() {
+    return new ParallelContextImpl();
```
A new instance on every call?

jovany-wang replied:
Fixed.

kfstorm commented on java/runtime/src/main/java/io/ray/runtime/utils/parallelactor/ParallelActorExecutorImpl.java:
```diff
@@ -0,0 +1,59 @@
+
+  public Object execute(int instanceIndex, JavaFunctionDescriptor functionDescriptor,  Object[] args) {
+
+    functionManager = ((RayRuntimeInternal) Ray.internal()).getFunctionManager();
```
Remove this line because `functionManager` is already initialized.

kfstorm commented on java/runtime/src/main/java/io/ray/runtime/utils/parallelactor/ParallelActorExecutorImpl.java:
```diff
@@ -0,0 +1,59 @@
+    functionManager = ((RayRuntimeInternal) Ray.internal()).getFunctionManager();
+    RayFunction func = functionManager.getFunction(Ray.getRuntimeContext().getCurrentJobId(), functionDescriptor);
+    try {
+      if (instances.get(instanceIndex) == null) {
```
In which case it is null? Shouldn't we check if the index is in range instead?

jovany-wang replied:
Used check instead.

kfstorm commented on java/runtime/src/main/java/io/ray/runtime/utils/parallelactor/ParallelActorExecutorImpl.java:
```diff
@@ -0,0 +1,59 @@
+
+      return func.getMethod().invoke(instances.get(instanceIndex), args);
+    } catch (Throwable a) {
+      throw new RuntimeException(a);
```
Maybe we should not catch and wrap exceptions here.

kfstorm commented on java/api/src/main/java/io/ray/api/utils/parallelactor/ParallelActorCreator.java:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  public ParallelActor<A> remote() {
+    ConcurrencyGroup[] concurrencyGroups = new ConcurrencyGroup[parallelNum];
+    for (int i = 0; i < parallelNum; ++i) {
+      concurrencyGroups[i] = new ConcurrencyGroupBuilder<ParallelActorExecutor>()
+        .setName(String.format("PARALLEL_INSTANCE_%d", i))
+        .setMaxConcurrency(1)
+        .build();
+    }
```
I believe this logic should be put in `createParallelActorExecutor`.

kfstorm commented on java/api/src/main/java/io/ray/api/utils/parallelactor/ParallelActorExecutor.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.util.List;
+
+public class ParallelActorExecutor {
```
Is it OK to put it into the runtime module?

jovany-wang replied:
If we hide the implementation of `ParallelActor` to runtime module, `ParallelActorExecutor` could be moved into runtime module as well.

kfstorm commented on java/api/src/main/java/io/ray/api/utils/parallelactor/ParallelActor.java:
```diff
@@ -0,0 +1,46 @@
+
+  private ParallelStrategy strategy;
+  private int parallelNum;
+  private int lastIndex = -1;
```
`-1` may not be a good initial value.

jovany-wang replied:
Hum, what's your suggestion for that?

kfstorm replied:
I'd say `lastIndex` is only used for the round-robin strategy. The always-first strategy doesn't utilize this field. It perfectly makes sense if there's a class to describe the current state (different implementations based on strategy) and move the `lastIndex` to the round-robin implementation. This is also more extensible if we consider more potential strategies in the future. (Note that `ParallelActor` is in the API package.)

jovany-wang replied:
Got your point.
I made the `Strategy` as an interface in API module, then `AlwaysFirstStrategy` and `RoundRobinStrategy` are in runtime module.

kfstorm commented on java/api/src/main/java/io/ray/api/utils/parallelactor/Parallel.java:
```diff
@@ -0,0 +1,6 @@
+package io.ray.api.utils.parallelactor;
```
`io.ray.api.utils.parallelactor` or `io.ray.api.parallelactor`?

kfstorm commented:
@jovany-wang Nice PR! Is there anything important not implemented yet?

jovany-wang commented:
> @jovany-wang Nice PR! Is there anything important not implemented yet?

Yes. We haven't implemented the invocations on an instance yet. I'm going to be for that ASAP.
```java

    ParallelInstance<A> instance = actor.getInstance(/*index=*/2);
    instance.task(A::f1, 100, 200).remote();   // Executed in instance 2
```

jovany-wang commented:
> > @jovany-wang Nice PR! Is there anything important not implemented yet?
> 
> Yes. We haven't implemented the invocations on an instance yet. I'm going to be for that ASAP.
> 
> ```java
>     ParallelInstance<A> instance = actor.getInstance(/*index=*/2);
>     instance.task(A::f1, 100, 200).remote();   // Executed in instance 2
> ```

This has been done and it's ready for review now. All major functionalities are implemented. I'm going to add full tests and do minor changes.

raulchen commented:
Can you update the PR description? Please add the new API as well.

raulchen commented on java/api/src/main/java/io/ray/api/parallelactor/ParallelActor.java:
```diff
@@ -0,0 +1,12 @@
+
+  ParallelStrategyInterface getStrategy();
+
+  ActorHandle<? extends ParallelActorExecutor> getExecutor();
```
Can we hide this method from the API?

raulchen commented on java/api/src/main/java/io/ray/api/parallelactor/ParallelActorCreator.java:
```diff
@@ -0,0 +1,34 @@
+    this.strategy = ctx.createParallelStrategy(ParallelStrategy.ALWAYS_FIRST, 1);
+  }
+
+  public ParallelActorCreator<A> setParallels(ParallelStrategy strategy, int parallelNum) {
```
1. We shouldn't use an enum `ParallelStrategy` as the parameter. Because then it's not possible for users to write their own strategy. Why not just use `ParallelStrategyInterface` as the parameter? 
2. better separate this into 2 methods `setStrategy` and `setParallelism`

jovany-wang replied:
I updated to
```java
Parallel.actor(A::new).setStrategy(new RoundRobinStrategy(/*parallelNum=*/10)).remote();
```

kfstorm replied:
@jovany-wang I don't think it's reasonable to put the parallelism info into the strategy class. I think `Parallel.actor(A::new).setParallelism(10).setStrategy(new RoundRobinStrategy()).remote()` is much better, because the parallelism option should not be a property of a strategy. And in this case, a user can omit `setStrategy` if one wants to use the default strategy. Still, `setParallelism` should be mandatory because the default value of 1 doesn't make sense.

jovany-wang replied:
All stuffs about strategy were removed.

raulchen commented on java/api/src/main/java/io/ray/api/parallelactor/ParallelActorCreator.java:
```diff
@@ -0,0 +1,34 @@
+
+  public ParallelActor<A> remote() {
+    ParallelContext ctx = Ray.internal().getParallelContext();
+    return ctx.createParallelActorExecutor(strategy, this.func);
```
Can we avoid exposing `ParallelContext` to the api module? Just do `Ray.internal().createParallelActor` here?

jovany-wang replied:
A ParalleContext could collect all helper methods:
![image](https://user-images.githubusercontent.com/19473861/163318836-d0b70a8d-7d81-4c5a-b12b-e42811b5c571.png)

raulchen commented on java/runtime/src/main/java/io/ray/runtime/utils/parallelactor/strategy/AlwaysFirstStrategy.java:
```diff
@@ -0,0 +1,28 @@
+import io.ray.api.parallelactor.ParallelStrategy;
+import io.ray.api.parallelactor.ParallelStrategyInterface;
+
+public class AlwaysFirstStrategy implements ParallelStrategyInterface {
```
This strategy looks weird. In which case would someone want to always call the first instance? If so, why bother creating multiple instances?
BTW, a random strategy would make sense.

kfstorm commented on java/test/src/main/java/io/ray/test/ParallelActorTest.java:
```diff
@@ -0,0 +1,80 @@
+    ObjectRef<Integer> obj0 = actor.task(A::incr, 1000000).remote(); // 调用的是instance 0
+    ObjectRef<Integer> obj1 = actor.task(A::incr, 1000000).remote(); // 调用的是instance 1
+    ObjectRef<Integer> obj2 = actor.task(A::incr, 1000000).remote(); // 调用的是instance 2
+    Assert.assertEquals(1000000, (int) obj0.get());
+    Assert.assertEquals(1000000, (int) obj1.get());
+    Assert.assertEquals(1000000, (int) obj2.get());
```
How about we pass different arguments to actors and expect different return values?

kfstorm commented on java/test/src/main/java/io/ray/test/ParallelActorTest.java:
```diff
@@ -0,0 +1,80 @@
+  /// 5. We used function manager.
+  public void testF() {
+    ParallelActor<A> actor =
+        Parallel.actor(A::new).setParallels(ParallelStrategy.ROUND_ROBIN, 10).remote();
```
I think you missed the test case for different strategies.

kfstorm commented on java/api/src/main/java/io/ray/api/parallelactor/VoidParallelActorTaskCaller.java:
```diff
@@ -0,0 +1,14 @@
+public class VoidParallelActorTaskCaller {
+
+  public VoidParallelActorTaskCaller(ParallelActor actor, RayFuncVoid func, Object[] args) {
+    // TODO: Fill in
```
Missing implemenation?

kfstorm commented on java/api/src/main/java/io/ray/api/parallelactor/ParallelActorExecutor.java:
```diff
@@ -0,0 +1,3 @@
+package io.ray.api.parallelactor;
+
+public class ParallelActorExecutor {}
```
What's the use of exposing a base class in the api module?

jovany-wang replied:
Removed

kfstorm commented on java/runtime/src/main/java/io/ray/runtime/utils/parallelactor/ParallelActorExecutorImpl.java:
```diff
@@ -0,0 +1,53 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ParallelActorExecutorImpl extends ParallelActorExecutor {
```
Is it possible to implement this feature without a dedicated actor class? One drawback I can think of is that all actor instances will have the name `ParallelActorExecutorImpl` on Ray dashboard, and it makes debugging and monitoring harder.

jovany-wang replied:
How about we add an actor title field and the it's default the class name by default we also could overwrite it when it's necessary, like:
```class
class A implements TitlableActor {
    @Override
    String getTitle() {
        // ...
    }
}
```

kfstorm replied:
Could you explain the possibility of implementing this without a special actor class? If it's not possible, why?

kfstorm replied:
In my understanding, the `ParallelActor` class is just a router, everything that happened under the hood should be the same compared to creating actors with a for loop.

jovany-wang replied:
I don't think it's easy to make if we don't change the Ray core code. or you could give an initial thought.

kfstorm replied:
Sorry that I misunderstood the purpose of this PR. The PR is to provide an alternative option for running multiple actor instances in a Java worker process, and the eventual goal is to remove the original multiple-CoreWorker-instance implementation.

Could you clarify this in the PR description?

For the actor naming issue on the dashboard, do you think automatically generating a unique actor name with the actor class name included is possible? e.g. When a user invokes `Parallel.actor(MyActor::new).setStrategy(new RoundRobinStrategy(10)).remote()`, the actor shows on the dashboard will have actor name `MyActor[10]-71a4ef`, where `71a4ef` is a random HEX string.

kfstorm replied:
We can put the auto naming feature into a later PR.

kfstorm reviewed (commented):
I'm wondering how users can pass a `ParallelActor` instance to other workers. We already support passing `ActorHandle` instances. Is it OK to just serialize the `ParallelActor` object? Do we need to make sure the state of `ParallelStrategyInterface` is erased duration serialization?

jovany-wang commented:
> I'm wondering how users can pass a ParallelActor instance to other workers. We already support passing ActorHandle instances. Is it OK to just serialize the ParallelActor object? Do we need to make sure the state of ParallelStrategyInterface is erased duration serialization?

In principle, I believe we have 2 important things for that:
1. we should make ParallelActor extend from ActorHandle to fit all APIs.
2. Run all existing tests on ParallelActor. (This is more important than the 1st action.)

But I don't think we need to make them in this PR. I prefer let this PR finishes the functionalities of ParallelActor.

jovany-wang commented on java/test/src/main/java/io/ray/test/ParallelActorTest.java:
```diff
@@ -0,0 +1,80 @@
+    ParallelActor<A> actor =
+        Parallel.actor(A::new).setParallels(ParallelStrategy.ROUND_ROBIN, 10).remote();
+
+    /// TODO: Why we set the delta so large? since it's cast to short otherwise.
```
Yes.

jovany-wang commented:
> I'm wondering how users can pass a ParallelActor instance to other workers. We already support passing ActorHandle instances. Is it OK to just serialize the ParallelActor object? Do we need to make sure the state of ParallelStrategyInterface is erased duration serialization?

I added the implementation to pass a parallel actor. It serializes actor handle and the necessary fields. 

For the strategy, I added a `void reset()` interface for `ParallellStrategyInterface` which will be invoked once deserialize a strategy.

jovany-wang commented:
@kfstorm @raulchen Feel free to take another look.

kfstorm commented on java/test/src/main/java/io/ray/test/ParallelActorTest.java:
```diff
@@ -0,0 +1,140 @@
+    /// TODO: Why we set the delta so large? because it's cast to short otherwise.
+    ObjectRef<Integer> obj0 = actor.task(A::incr, 1000000).remote(); // in instance 0
+    ObjectRef<Integer> obj1 = actor.task(A::incr, 2000000).remote(); // in instance 1
+    ObjectRef<Integer> obj2 = actor.task(A::incr, 3000000).remote(); // in instance 2
```
Can we add a getPid method to the actor class and verify that all of them are in the same worker process?

kfstorm commented on java/api/src/main/java/io/ray/api/parallelactor/ParallelActorHandle.java:
```diff
@@ -0,0 +1,25 @@
+ * <p>Every thread holds an instance to `A`, and the methods will be invoked on the instance which
+ * is calculated according to the parallel strategy.
+ *
+ * <p>The parallel strategy are only working on the caller side, which means it doesn't be
```
Remove strategy-related comments.

kfstorm commented on java/api/src/main/java/io/ray/api/parallelactor/ParallelActorHandle.java:
```diff
@@ -0,0 +1,25 @@
+  ParallelInstance<A> getInstance(int index);
+
+  /** Get the parallel number of this parallel actor. */
+  int getParallelNum();
```
Maybe `getParallelism` is better because we use `setParallelism` when creating it.

kfstorm commented on java/api/src/main/java/io/ray/api/parallelactor/ParallelActorCreator.java:
```diff
@@ -0,0 +1,28 @@
+    this.args = args;
+  }
+
+  public ParallelActorCreator<A> setParallelism(int parallelNum) {
```
rename `parallelNum` to `parallelism`.

kfstorm reviewed (commented):
I suggest to do the following renames:

ParallelActorCall -> ActorCall
ParallelCall -> Call
ParallelContext -> ParallelActorContext
ParallelInstance -> ParallelActorInstance

Rename 1) and 2) because the current naming is confusing. Since they are not really user facing classes, I think it's OK to remove the `ParallelActor` prefix.

Rename 3) and 4) because the prefix should be `ParallelActor`, not `Parallel`.

kfstorm commented on java/api/src/main/java/io/ray/api/parallelactor/ParallelActorHandle.java:
```diff
@@ -0,0 +1,22 @@
+  /** Get an execution instance of the parallel actor by the given index. */
+  ParallelActorInstance<A> getInstance(int index);
+
+  /** Get the parallel number of this parallel actor. */
```
```suggestion
  /** Get the parallelism of this parallel actor. */
```

Maybe `parallelism` is not a good word.

kfstorm commented on java/api/src/main/java/io/ray/api/parallelactor/ParallelActorInstance.java:
```diff
@@ -0,0 +1,21 @@
+    return parallelActorHandle;
+  }
+
+  int getIndex() {
```
```suggestion
  int getInstanceId() {
```

kfstorm commented on java/api/src/main/java/io/ray/api/parallelactor/ParallelActorInstance.java:
```diff
@@ -0,0 +1,21 @@
+
+  private ParallelActorHandle<A> parallelActorHandle;
+
+  private int currentIndex;
```
```suggestion
  private int instanceId;
```

kfstorm commented on java/api/src/main/java/io/ray/api/parallelactor/ParallelActorInstance.java:
```diff
@@ -0,0 +1,21 @@
+
+  private int currentIndex;
+
+  public ParallelActorInstance(ParallelActorHandle<A> parallelActorHandle, int index) {
```
```suggestion
  public ParallelActorInstance(ParallelActorHandle<A> parallelActorHandle, int instanceId) {
```

kfstorm commented on java/api/src/main/java/io/ray/api/runtime/RayRuntime.java:
```diff
@@ -286,4 +287,7 @@ <T> ActorHandle<T> createActor(
   /** Create runtime env instance at runtime. */
   RuntimeEnv createRuntimeEnv(Map<String, String> envVars);
+
+  /// Get the parallel context at runtime.
+  ParallelActorContext getParallelContext();
```
```suggestion
  /// Get the parallel actor context at runtime.
  ParallelActorContext getParallelActorContext();
```

kfstorm commented on java/runtime/src/main/java/io/ray/runtime/AbstractRayRuntime.java:
```diff
@@ -288,6 +292,11 @@ public List<ConcurrencyGroup> extractConcurrencyGroups(RayFuncR<?> actorConstruc
 
+  @Override
+  public ParallelActorContext getParallelContext() {
+    return parallelContextImpl;
```
```suggestion
    return parallelActorContextImpl;
```

kfstorm commented on java/runtime/src/main/java/io/ray/runtime/utils/parallelactor/ParallelActorContextImpl.java:
```diff
@@ -0,0 +1,59 @@
+
+  @Override
+  public <A, R> ObjectRef<R> submitTask(
+      ParallelActorHandle<A> parallelActorHandle, int instanceIndex, RayFunc func, Object[] args) {
```
```suggestion
      ParallelActorHandle<A> parallelActorHandle, int instanceId, RayFunc func, Object[] args) {
```

kfstorm commented on java/runtime/src/main/java/io/ray/runtime/utils/parallelactor/ParallelActorHandleImpl.java:
```diff
@@ -0,0 +1,55 @@
+    return parallelExecutorHandle;
+  }
+
+  @Override
+  public void writeExternal(ObjectOutput out) throws IOException {
+    out.writeObject(parallelExecutorHandle);
+    out.writeInt(parallelism);
+  }
+
+  @Override
+  public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
+    this.parallelExecutorHandle = (ActorHandle<ParallelActorExecutorImpl>) in.readObject();
+    this.parallelism = in.readInt();
+  }
```
We can remove these and just implements `Serializable`.

kfstorm commented on java/test/src/main/java/io/ray/test/ParallelActorTest.java:
```diff
@@ -0,0 +1,128 @@
+      int res = instance.task(A::add, 100000, 200000).remote().get(); // Executed in instance 2
+      Assert.assertEquals(res, 300000);
+
+      instance = actor.getInstance(/*index=*/ 3);
```
```suggestion
      instance = actor.getInstance(/*instanceId=*/ 3);
```

kfstorm commented on java/runtime/src/main/java/io/ray/runtime/util/generator/ParallelActorCallGenerator.java:
```diff
@@ -0,0 +1,334 @@
+    newLine("");
+
+    newLine("/**");
+    newLine(" * This class provides type-safe interfaces for `Ray.call` and `Ray.createActor`.");
```
```suggestion
    newLine(" * This class provides type-safe interfaces for `ParallelActor.actor`.");
```

kfstorm commented on java/runtime/src/main/java/io/ray/runtime/util/generator/ParallelActorCallGenerator.java:
```diff
@@ -0,0 +1,334 @@
+    newLine("/**");
+    newLine(" * This class provides type-safe interfaces for `Ray.call` and `Ray.createActor`.");
+    newLine(" **/");
+    newLine("class ParallelCall {");
```
```suggestion
    newLine("class Call {");
```

kfstorm commented on java/api/src/main/java/io/ray/api/parallelactor/ParallelActorHandle.java:
```diff
@@ -0,0 +1,22 @@
+ */
+public interface ParallelActorHandle<A> {
+
+  /** Get an execution instance of the parallel actor by the given index. */
```
```suggestion
  /** Get an execution instance of the parallel actor by the given instance ID. */
```

kfstorm commented on java/api/src/main/java/io/ray/api/parallelactor/ParallelActorHandle.java:
```diff
@@ -0,0 +1,22 @@
+import io.ray.api.ActorHandle;
+
+/**
+ * A `ParallelActorHandle` could be used for creating multiple execution instances in different
+ * threads, to execute the methods.
+ *
+ * <p>Every thread holds an instance to `A`, and the methods will be invoked on the instance in
+ * different threads.
```
I think this comment should be on `ParallelActor`, not `ParallelActorHandle`.
